File: python/bulldb/query.py
from typing import Any, Dict, List, Optional, Tuple, Type, Union

class ASTNode:
    pass

class ValueNode(ASTNode):
    def __init__(self, value: Any):
        self.value = value

class ColumnNode(ASTNode):
    def __init__(self, name: str):
        self.name = name

class BinaryOpNode(ASTNode):
    def __init__(self, left: ASTNode, op: str, right: ASTNode):
        self.left = left
        self.op = op
        self.right = right

class SelectAST(ASTNode):
    def __init__(self):
        self.table: str = ""
        self.fields: List[str] = []
        self.filters: Optional[ASTNode] = None
        self.joins: List[Tuple[str, str, str]] = [] # target_table, on_left, on_right
        self.order_by: List[str] = []
        self.limit: Optional[int] = None
        self.offset: Optional[int] = None
        self.vector_search: Optional[Tuple[str, List[float], int]] = None # field, query_vector, limit
        self.ctes: List[Tuple[str, 'SelectAST']] = []

class QueryCompiler:
    @staticmethod
    def compile_to_sql(ast: SelectAST) -> Tuple[str, tuple]:
        params = []
        
        # Compile CTEs
        cte_parts = []
        for cte_name, cte_ast in ast.ctes:
            sub_sql, sub_params = QueryCompiler.compile_to_sql(cte_ast)
            cte_parts.append(f"{cte_name} AS ({sub_sql})")
            params.extend(sub_params)
        cte_prefix = f"WITH {', '.join(cte_parts)} " if cte_parts else ""

        # Fields
        fields_str = ", ".join(ast.fields) if ast.fields else "*"
        
        # Base Query
        sql = f"{cte_prefix}SELECT {fields_str} FROM {ast.table}"
        
        # Joins
        for target, left, right in ast.joins:
            sql += f" JOIN {target} ON {ast.table}.{left} = {target}.{right}"
            
        # Where filters
        if ast.filters:
            where_sql, where_params = QueryCompiler._compile_filter_sql(ast.filters)
            sql += f" WHERE {where_sql}"
            params.extend(where_params)

        # Vector search fallback ordering (e.g. pgvector)
        if ast.vector_search:
            field, vector, limit = ast.vector_search
            # Postgres pgvector operator for cosine distance is <=>
            sql += f" ORDER BY {field} <=> %s"
            params.append(str(vector))
            if not ast.limit:
                ast.limit = limit

        # Order by
        if ast.order_by:
            sql += f" ORDER BY {', '.join(ast.order_by)}"
            
        # Limit / Offset
        if ast.limit is not None:
            sql += f" LIMIT {ast.limit}"
        if ast.offset is not None:
            sql += f" OFFSET {ast.offset}"
            
        return sql, tuple(params)

    @staticmethod
    def _compile_filter_sql(node: ASTNode) -> Tuple[str, list]:
        if isinstance(node, ValueNode):
            return "?", [node.value]
        if isinstance(node, ColumnNode):
            return node.name, []
        if isinstance(node, BinaryOpNode):
            left_sql, left_params = QueryCompiler._compile_filter_sql(node.left)
            right_sql, right_params = QueryCompiler._compile_filter_sql(node.right)
            return f"({left_sql} {node.op} {right_sql})", left_params + right_params
        return "", []

    @staticmethod
    def compile_to_mongo(ast: SelectAST) -> Tuple[Dict[str, Any], Dict[str, Any], Optional[int]]:
        # Returns (filter_query, projection, limit)
        filters = {}
        if ast.filters:
            filters = QueryCompiler._compile_filter_mongo(ast.filters)
            
        projection = {f: 1 for f in ast.fields} if ast.fields else {}
        return filters, projection, ast.limit

    @staticmethod
    def _compile_filter_mongo(node: ASTNode) -> Dict[str, Any]:
        if isinstance(node, BinaryOpNode):
            # Parse field and value
            left = node.left
            right = node.right
            if isinstance(left, ColumnNode) and isinstance(right, ValueNode):
                field = left.name
                val = right.value
                op_map = {"=": "$eq", "!=": "$ne", ">": "$gt", "<": "$lt", ">=": "$gte", "<=": "$lte"}
                mongo_op = op_map.get(node.op, "$eq")
                return {field: {mongo_op: val}}
            # Compound AND/OR
            if node.op.upper() in ("AND", "OR"):
                left_mongo = QueryCompiler._compile_filter_mongo(left)
                right_mongo = QueryCompiler._compile_filter_mongo(right)
                key = "$and" if node.op.upper() == "AND" else "$or"
                return {key: [left_mongo, right_mongo]}
        return {}

    @staticmethod
    def compile_to_cypher(ast: SelectAST) -> Tuple[str, dict]:
        # Returns cypher string and params dict
        params = {}
        match_pattern = f"(n:{ast.table})"
        where_clauses = []
        
        if ast.filters:
            cypher_filter, cypher_params = QueryCompiler._compile_filter_cypher(ast.filters)
            where_clauses.append(cypher_filter)
            params.update(cypher_params)
            
        where_str = f" WHERE {', '.join(where_clauses)}" if where_clauses else ""
        return_fields = ", ".join([f"n.{f}" for f in ast.fields]) if ast.fields else "n"
        
        cypher = f"MATCH {match_pattern}{where_str} RETURN {return_fields}"
        if ast.limit:
            cypher += f" LIMIT {ast.limit}"
            
        return cypher, params

    @staticmethod
    def _compile_filter_cypher(node: ASTNode) -> Tuple[str, dict]:
        if isinstance(node, BinaryOpNode):
            left = node.left
            right = node.right
            if isinstance(left, ColumnNode) and isinstance(right, ValueNode):
                field = left.name
                param_name = f"val_{field}"
                op = node.op
                if op == "=": op = "=="
                return f"n.{field} {op} ${param_name}", {param_name: right.value}
        return "", {}

    @staticmethod
    def compile_to_elasticsearch(ast: SelectAST) -> Tuple[Dict[str, Any], int]:
        query_body = {"query": {"match_all": {}}}
        if ast.filters:
            query_body = {"query": QueryCompiler._compile_filter_es(ast.filters)}
        if ast.fields:
            query_body["_source"] = ast.fields
        return query_body, ast.limit or 10

    @staticmethod
    def _compile_filter_es(node: ASTNode) -> Dict[str, Any]:
        if isinstance(node, BinaryOpNode):
            left = node.left
            right = node.right
            if isinstance(left, ColumnNode) and isinstance(right, ValueNode):
                field = left.name
                val = right.value
                if node.op == "=":
                    return {"term": {field: val}}
                elif node.op == "!=":
                    return {"bool": {"must_not": [{"term": {field: val}}]}}
                elif node.op in (">", "<", ">=", "<="):
                    es_op = {">": "gt", "<": "lt", ">=": "gte", "<=": "lte"}[node.op]
                    return {"range": {field: {es_op: val}}}
            if node.op.upper() == "AND":
                return {"bool": {"must": [QueryCompiler._compile_filter_es(left), QueryCompiler._compile_filter_es(right)]}}
            if node.op.upper() == "OR":
                return {"bool": {"should": [QueryCompiler._compile_filter_es(left), QueryCompiler._compile_filter_es(right)]}}
        return {"match_all": {}}

    @staticmethod
    def compile_to_vector_query(ast: SelectAST) -> Tuple[Optional[List[float]], Dict[str, Any], int]:
        filters = {}
        if ast.filters:
            filters = QueryCompiler._compile_filter_vector(ast.filters)
        vector = None
        limit = ast.limit or 5
        if ast.vector_search:
            _, vector, limit = ast.vector_search
        return vector, filters, limit

    @staticmethod
    def _compile_filter_vector(node: ASTNode) -> Dict[str, Any]:
        if isinstance(node, BinaryOpNode):
            left = node.left
            right = node.right
            if isinstance(left, ColumnNode) and isinstance(right, ValueNode):
                return {left.name: right.value}
            res = {}
            res.update(QueryCompiler._compile_filter_vector(left))
            res.update(QueryCompiler._compile_filter_vector(right))
            return res
        return {}

class QueryBuilder:
    def __init__(self, model_class: Type):
        self.model_class = model_class
        self.ast = SelectAST()
        self.ast.table = model_class._table_name
        self.ast.fields = list(model_class._fields.keys())

    def select(self, *fields: str) -> "QueryBuilder":
        self.ast.fields = list(fields)
        return self

    def where(self, field: str, op: str, value: Any) -> "QueryBuilder":
        new_filter = BinaryOpNode(ColumnNode(field), op, ValueNode(value))
        if self.ast.filters:
            self.ast.filters = BinaryOpNode(self.ast.filters, "AND", new_filter)
        else:
            self.ast.filters = new_filter
        return self

    def and_where(self, field: str, op: str, value: Any) -> "QueryBuilder":
        return self.where(field, op, value)

    def or_where(self, field: str, op: str, value: Any) -> "QueryBuilder":
        new_filter = BinaryOpNode(ColumnNode(field), op, ValueNode(value))
        if self.ast.filters:
            self.ast.filters = BinaryOpNode(self.ast.filters, "OR", new_filter)
        else:
            self.ast.filters = new_filter
        return self

    def join(self, target_table: str, on_left: str, on_right: str) -> "QueryBuilder":
        self.ast.joins.append((target_table, on_left, on_right))
        return self

    def order_by(self, field: str, direction: str = "ASC") -> "QueryBuilder":
        suffix = f" {direction}" if direction.upper() == "DESC" else ""
        self.ast.order_by.append(f"{field}{suffix}")
        return self

    def limit(self, limit_val: int) -> "QueryBuilder":
        self.ast.limit = limit_val
        return self

    def offset(self, offset_val: int) -> "QueryBuilder":
        self.ast.offset = offset_val
        return self

    def vector_search(self, field: str, query_vector: List[float], limit: int = 5) -> "QueryBuilder":
        self.ast.vector_search = (field, query_vector, limit)
        return self

    def with_cte(self, cte_name: str, subquery: "QueryBuilder") -> "QueryBuilder":
        self.ast.ctes.append((cte_name, subquery.ast))
        return self

    async def execute(self) -> List[Any]:
        db = self.model_class._db
        if not db:
            raise ValueError("No database client registered. Connect to MultiDatabase first.")

        driver = db.get_route(self.ast.table)
        dialect = driver.name

        from .security import SecurityEngine
        SecurityEngine.inject_rls(self.ast)

        results = []
        if "mongo" in dialect:
            filters, projection, limit = QueryCompiler.compile_to_mongo(self.ast)
            mongo_results = await driver.execute_mongo_find(self.ast.table, filters, projection, limit)
            results = mongo_results
        elif "elasticsearch" in dialect:
            body, limit = QueryCompiler.compile_to_elasticsearch(self.ast)
            results = await driver.execute_search(self.ast.table, body, limit)
        elif "vector" in dialect or "chroma" in dialect or "pinecone" in dialect:
            vector, filters, limit = QueryCompiler.compile_to_vector_query(self.ast)
            results = await driver.execute_vector_search(self.ast.table, vector, filters, limit)
        elif "neo4j" in dialect:
            cypher, params = QueryCompiler.compile_to_cypher(self.ast)
            results = await driver.execute(cypher, tuple(params.values()))
        else:
            sql, params = QueryCompiler.compile_to_sql(self.ast)
            results = await driver.execute(sql, params)

        return [self.model_class(**res) for res in results]

File: typescript/src/query.ts
import { ModelMetadataRegistry } from "./models";
import { SecurityEngine } from "./security";

export abstract class ASTNode {}

export class ValueNode extends ASTNode {
  constructor(public value: any) { super(); }
}

export class ColumnNode extends ASTNode {
  constructor(public name: string) { super(); }
}

export class BinaryOpNode extends ASTNode {
  constructor(public left: ASTNode, public op: string, public right: ASTNode) { super(); }
}

export class SelectAST extends ASTNode {
  public table = "";
  public fields: string[] = [];
  public filters: ASTNode | null = null;
  public joins: [string, string, string][] = []; // target_table, on_left, on_right
  public orderBy: string[] = [];
  public limit: number | null = null;
  public offset: number | null = null;
  public vectorSearch: [string, number[], number] | null = null; // field, query_vector, limit
  public ctes: [string, SelectAST][] = [];
}

export class QueryCompiler {
  static compileToSql(ast: SelectAST): [string, any[]] {
    const params: any[] = [];
    
    // 1. CTE Compilation
    let ctePrefix = "";
    if (ast.ctes.length > 0) {
      const parts = ast.ctes.map(([name, cteAst]) => {
        const [subSql, subParams] = this.compileToSql(cteAst);
        params.push(...subParams);
        return `${name} AS (${subSql})`;
      });
      ctePrefix = `WITH ${parts.join(", ")} `;
    }

    const fieldsStr = ast.fields.length > 0 ? ast.fields.join(", ") : "*";
    let sql = `${ctePrefix}SELECT ${fieldsStr} FROM ${ast.table}`;

    // Joins
    for (const [target, left, right] of ast.joins) {
      sql += ` JOIN ${target} ON ${ast.table}.${left} = ${target}.${right}`;
    }

    // Where filters
    if (ast.filters) {
      const [whereSql, whereParams] = this.compileFilterSql(ast.filters);
      sql += ` WHERE ${whereSql}`;
      params.push(...whereParams);
    }

    // Vector query additions
    if (ast.vectorSearch) {
      const [field, vector, limit] = ast.vectorSearch;
      sql += ` ORDER BY ${field} <=> ?`;
      params.push(JSON.stringify(vector));
      if (ast.limit === null) {
        ast.limit = limit;
      }
    }

    // Order By
    if (ast.orderBy.length > 0) {
      sql += ` ORDER BY ${ast.orderBy.join(", ")}`;
    }

    // Limit / Offset
    if (ast.limit !== null) sql += ` LIMIT ${ast.limit}`;
    if (ast.offset !== null) sql += ` OFFSET ${ast.offset}`;

    return [sql, params];
  }

  private static compileFilterSql(node: ASTNode): [string, any[]] {
    if (node instanceof ValueNode) {
      return ["?", [node.value]];
    }
    if (node instanceof ColumnNode) {
      return [node.name, []];
    }
    if (node instanceof BinaryOpNode) {
      const [leftSql, leftParams] = this.compileFilterSql(node.left);
      const [rightSql, rightParams] = this.compileFilterSql(node.right);
      return [`(${leftSql} ${node.op} ${rightSql})`, [...leftParams, ...rightParams]];
    }
    return ["", []];
  }

  static compileToMongo(ast: SelectAST): [Record<string, any>, Record<string, any>, number | null] {
    const filters = ast.filters ? this.compileFilterMongo(ast.filters) : {};
    const projection: Record<string, any> = {};
    for (const field of ast.fields) {
      projection[field] = 1;
    }
    return [filters, projection, ast.limit];
  }

  private static compileFilterMongo(node: ASTNode): Record<string, any> {
    if (node instanceof BinaryOpNode) {
      const { left, op, right } = node;
      if (left instanceof ColumnNode && right instanceof ValueNode) {
        const opMap: Record<string, string> = {
          "=": "$eq",
          "!=": "$ne",
          ">": "$gt",
          "<": "$lt",
          ">=": "$gte",
          "<=": "$lte"
        };
        const mongoOp = opMap[op] || "$eq";
        return { [left.name]: { [mongoOp]: right.value } };
      }
      if (["AND", "OR"].includes(op.toUpperCase())) {
        const key = op.toUpperCase() === "AND" ? "$and" : "$or";
        return { [key]: [this.compileFilterMongo(left), this.compileFilterMongo(right)] };
      }
    }
    return {};
  }

  static compileToCypher(ast: SelectAST): [string, Record<string, any>] {
    const params: Record<string, any> = {};
    const match = `(n:${ast.table})`;
    let whereStr = "";
    if (ast.filters) {
      const [cyFilter, cyParams] = this.compileFilterCypher(ast.filters);
      whereStr = ` WHERE ${cyFilter}`;
      Object.assign(params, cyParams);
    }
    const returnFields = ast.fields.length > 0 ? ast.fields.map(f => `n.${f}`).join(", ") : "n";
    let cypher = `MATCH ${match}${whereStr} RETURN ${returnFields}`;
    if (ast.limit !== null) {
      cypher += ` LIMIT ${ast.limit}`;
    }
    return [cypher, params];
  }

  private static compileFilterCypher(node: ASTNode): [string, Record<string, any>] {
    if (node instanceof BinaryOpNode) {
      const { left, op, right } = node;
      if (left instanceof ColumnNode && right instanceof ValueNode) {
        const paramName = `val_${left.name}`;
        const cyOp = op === "=" ? "==" : op;
        return [`n.${left.name} ${cyOp} $${paramName}`, { [paramName]: right.value }];
      }
    }
    return ["", {}];
  }

  static compileToElasticsearch(ast: SelectAST): [Record<string, any>, number] {
    let queryBody: Record<string, any> = { query: { match_all: {} } };
    if (ast.filters) {
      queryBody = { query: this.compileFilterEs(ast.filters) };
    }
    if (ast.fields.length > 0) {
      queryBody._source = ast.fields;
    }
    return [queryBody, ast.limit || 10];
  }

  private static compileFilterEs(node: ASTNode): Record<string, any> {
    if (node instanceof BinaryOpNode) {
      const { left, op, right } = node;
      if (left instanceof ColumnNode && right instanceof ValueNode) {
        if (op === "=") {
          return { term: { [left.name]: right.value } };
        } else if (op === "!=") {
          return { bool: { must_not: [{ term: { [left.name]: right.value } }] } };
        } else if ([">", "<", ">=", "<="].includes(op)) {
          const esOp = { ">": "gt", "<": "lt", ">=": "gte", "<=": "lte" }[op] as string;
          return { range: { [left.name]: { [esOp]: right.value } } };
        }
      }
      if (op.toUpperCase() === "AND") {
        return { bool: { must: [this.compileFilterEs(left), this.compileFilterEs(right)] } };
      }
      if (op.toUpperCase() === "OR") {
        return { bool: { should: [this.compileFilterEs(left), this.compileFilterEs(right)] } };
      }
    }
    return { match_all: {} };
  }

  static compileToVectorQuery(ast: SelectAST): [number[] | null, Record<string, any>, number] {
    const filters = ast.filters ? this.compileFilterVector(ast.filters) : {};
    let vector: number[] | null = null;
    let limit = ast.limit || 5;
    if (ast.vectorSearch) {
      [, vector, limit] = ast.vectorSearch;
    }
    return [vector, filters, limit];
  }

  private static compileFilterVector(node: ASTNode): Record<string, any> {
    if (node instanceof BinaryOpNode) {
      const { left, op, right } = node;
      if (left instanceof ColumnNode && right instanceof ValueNode) {
        return { [left.name]: right.value };
      }
      const res: Record<string, any> = {};
      Object.assign(res, this.compileFilterVector(left));
      Object.assign(res, this.compileFilterVector(right));
      return res;
    }
    return {};
  }
}

export class QueryBuilder<T> {
  public ast = new SelectAST();

  constructor(private modelClass: any) {
    this.ast.table = ModelMetadataRegistry.getTableName(modelClass.name);
    const fields = ModelMetadataRegistry.getFields(modelClass.name);
    this.ast.fields = Array.from(fields.keys());
  }

  select(...fields: string[]): this {
    this.ast.fields = fields;
    return this;
  }

  where(field: string, op: string, value: any): this {
    const cond = new BinaryOpNode(new ColumnNode(field), op, new ValueNode(value));
    if (this.ast.filters) {
      this.ast.filters = new BinaryOpNode(this.ast.filters, "AND", cond);
    } else {
      this.ast.filters = cond;
    }
    return this;
  }

  andWhere(field: string, op: string, value: any): this {
    return this.where(field, op, value);
  }

  orWhere(field: string, op: string, value: any): this {
    const cond = new BinaryOpNode(new ColumnNode(field), op, new ValueNode(value));
    if (this.ast.filters) {
      this.ast.filters = new BinaryOpNode(this.ast.filters, "OR", cond);
    } else {
      this.ast.filters = cond;
    }
    return this;
  }

  join(targetTable: string, onLeft: string, onRight: string): this {
    this.ast.joins.push([targetTable, onLeft, onRight]);
    return this;
  }

  orderBy(field: string, direction: "ASC" | "DESC" = "ASC"): this {
    const dir = direction === "DESC" ? " DESC" : "";
    this.ast.orderBy.push(`${field}${dir}`);
    return this;
  }

  limit(limitVal: number): this {
    this.ast.limit = limitVal;
    return this;
  }

  offset(offsetVal: number): this {
    this.ast.offset = offsetVal;
    return this;
  }

  vectorSearch(field: string, queryVector: number[], limit = 5): this {
    this.ast.vectorSearch = [field, queryVector, limit];
    return this;
  }

  withCte(cteName: string, subquery: QueryBuilder<any>): this {
    this.ast.ctes.push([cteName, subquery.ast]);
    return this;
  }

  async execute(): Promise<T[]> {
    const db = this.modelClass.db;
    if (!db) {
      throw new Error("No database client registered. Call setDb() on model class first.");
    }

    const driver = db.getRoute(this.ast.table);
    const dialect = driver.name;

    // Inject RLS parameters
    SecurityEngine.injectRls(this.ast);

    let results: any[] = [];
    if (dialect.includes("mongo")) {
      const [filters, projection, limit] = QueryCompiler.compileToMongo(this.ast);
      if (typeof (driver as any).executeMongoFind === "function") {
        results = await (driver as any).executeMongoFind(this.ast.table, filters, projection, limit);
      } else {
        results = await driver.execute(JSON.stringify({ filters, projection, limit }));
      }
    } else if (dialect.includes("elasticsearch")) {
      const [body, limit] = QueryCompiler.compileToElasticsearch(this.ast);
      if (typeof (driver as any).executeSearch === "function") {
        results = await (driver as any).executeSearch(this.ast.table, body, limit);
      } else {
        results = await driver.execute(JSON.stringify({ body, limit }));
      }
    } else if (dialect.includes("vector") || dialect.includes("chroma") || dialect.includes("pinecone")) {
      const [vector, filters, limit] = QueryCompiler.compileToVectorQuery(this.ast);
      if (typeof (driver as any).executeVectorSearch === "function") {
        results = await (driver as any).executeVectorSearch(this.ast.table, vector, filters, limit);
      } else {
        results = await driver.execute(JSON.stringify({ vector, filters, limit }));
      }
    } else if (dialect.includes("neo4j")) {
      const [cypher, params] = QueryCompiler.compileToCypher(this.ast);
      results = await driver.execute(cypher, Object.values(params));
    } else {
      // SQL fallback
      const [sql, params] = QueryCompiler.compileToSql(this.ast);
      results = await driver.execute(sql, params);
    }

    return results.map(row => new this.modelClass(row));
  }
}
